File: backend/models/News.js
const mongoose = require('mongoose');

const newsSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Заголовок новости обязателен'],
    trim: true,
    maxlength: [200, 'Заголовок не должен превышать 200 символов']
  },
  content: {
    type: String,
    required: [true, 'Содержание новости обязательно']
  },
  excerpt: {
    type: String,
    maxlength: [300, 'Краткое описание не должно превышать 300 символов']
  },
  thumbnail: {
    type: String
  },
  category: {
    type: String,
    enum: ['news', 'article', 'guide', 'event_report', 'research'],
    default: 'news'
  },
  tags: [String],
  author: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  status: {
    type: String,
    enum: ['draft', 'pending', 'approved', 'rejected'],
    default: 'pending'
  },
  moderator: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  moderationNote: String,
  publishedAt: Date,
  views: {
    type: Number,
    default: 0,
    min: 0
  },
  likes: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  comments: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    text: {
      type: String,
      required: true,
      maxlength: [500, 'Комментарий не должен превышать 500 символов']
    },
    createdAt: {
      type: Date,
      default: Date.now
    }
  }]
}, {
  timestamps: true
});

// Индексы
newsSchema.index({ status: 1, publishedAt: -1 });
newsSchema.index({ author: 1, status: 1 });
newsSchema.index({ category: 1, status: 1 });
newsSchema.index({ tags: 1 });

// Автоматически устанавливаем publishedAt при одобрении
newsSchema.pre('save', function(next) {
  if (this.isModified('status') && this.status === 'approved' && !this.publishedAt) {
    this.publishedAt = new Date();
  }
  next();
});

// Метод для лайка
newsSchema.methods.toggleLike = function(userId) {
  const index = this.likes.indexOf(userId);
  if (index === -1) {
    this.likes.push(userId);
  } else {
    this.likes.splice(index, 1);
  }
  return this.save();
};

// Метод для добавления комментария
newsSchema.methods.addComment = function(userId, text) {
  this.comments.push({ user: userId, text });
  return this.save();
};

// Метод для увеличения просмотров
newsSchema.methods.incrementViews = function() {
  this.views += 1;
  return this.save();
};

module.exports = mongoose.model('News', newsSchema);

File: frontend/lib/models/news.dart
class News {
  final String id;
  final String title;
  final String content;
  final String? excerpt;
  final String? thumbnail;
  final String category;
  final List<String> tags;
  final String authorId;
  final String? authorName;
  final String status;
  final DateTime? publishedAt;
  final int views;
  final List<String> likes;
  final List<NewsComment> comments;
  final DateTime createdAt;

  News({
    required this.id,
    required this.title,
    required this.content,
    this.excerpt,
    this.thumbnail,
    required this.category,
    required this.tags,
    required this.authorId,
    this.authorName,
    required this.status,
    this.publishedAt,
    required this.views,
    required this.likes,
    required this.comments,
    required this.createdAt,
  });

  factory News.fromJson(Map<String, dynamic> json) {
    return News(
      id: json['_id'] ?? '',
      title: json['title'] ?? '',
      content: json['content'] ?? '',
      excerpt: json['excerpt'],
      thumbnail: json['thumbnail'],
      category: json['category'] ?? '',
      tags: List<String>.from(json['tags'] ?? []),
      authorId: json['author'] != null
          ? (json['author'] is String ? json['author'] : json['author']['_id'] ?? '')
          : '',
      authorName: json['author'] != null && json['author'] is Map
          ? json['author']['username']
          : null,
      status: json['status'] ?? 'pending',
      publishedAt: json['publishedAt'] != null 
          ? DateTime.parse(json['publishedAt']) 
          : null,
      views: json['views'] ?? 0,
      likes: List<String>.from(json['likes'] ?? []),
      comments: (json['comments'] as List?)
          ?.map((c) => NewsComment.fromJson(c))
          .toList() ?? [],
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  bool get isApproved => status == 'approved';
  int get likesCount => likes.length;
  int get commentsCount => comments.length;
}

class NewsComment {
  final String userId;
  final String? username;
  final String text;
  final DateTime createdAt;

  NewsComment({
    required this.userId,
    this.username,
    required this.text,
    required this.createdAt,
  });

  factory NewsComment.fromJson(Map<String, dynamic> json) {
    return NewsComment(
      userId: json['user'] != null
          ? (json['user'] is String ? json['user'] : json['user']['_id'] ?? '')
          : '',
      username: json['user'] != null && json['user'] is Map
          ? json['user']['username']
          : null,
      text: json['text'] ?? '',
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
    );
  }
}

